% testing a loader for table data ;
clear;
setup;
dir_trunk = sprintf('/data/rangan/dir_bcc/dir_jamison');
dir_data = sprintf('%s/data_summary_20190730',dir_trunk);
%%%%%%%%;
flag_load=0;
if flag_load;
%%%%%%%%;
str_table_name = sprintf('%s/20161026_covariate_table.format.txt',dir_data);
disp(sprintf(' %% reading %s',str_table_name));
C_ = readtable(str_table_name);
disp(sprintf(' %% saving %s',str_table_name));
save(sprintf('%s/20161026_covariate_table.format.mat',dir_data),'C_');
%%%%%%%%;
str_table_name = sprintf('%s/EXON.Gx_50421.Cx_50.brain_table_merge_output.CPM.NotCtrl.NotEmpty.txt',dir_data);
disp(sprintf(' %% reading %s',str_table_name));
E_ = readtable(str_table_name);
disp(sprintf(' %% writing %s',str_table_name));
save(sprintf('%s/EXON.Gx_50421.Cx_50.brain_table_merge_output.CPM.NotCtrl.NotEmpty.mat',dir_data),'E_');
%%%%%%%%;
str_table_name = sprintf('%s/INTRON_INTERGENIC.Gx_25998.Cx_50.brain_table_merge_output.CPM.NotCtrl.NotEmpty.txt',dir_data);
disp(sprintf(' %% reading %s',str_table_name));
I_ = readtable(str_table_name);
disp(sprintf(' %% writing %s',str_table_name));
save(sprintf('%s/INTRON_INTERGENIC.Gx_25998.Cx_50.brain_table_merge_output.CPM.NotCtrl.NotEmpty.mat',dir_data),'I_');
%%%%%%%%;
end;%if flag_load;
%%%%%%%%;

flag_E = 1;
flag_I = 1;

%%%%%%%%;
load(sprintf('%s/EXON.Gx_50421.Cx_50.brain_table_merge_output.CPM.NotCtrl.NotEmpty.mat',dir_data),'E_');
load(sprintf('%s/INTRON_INTERGENIC.Gx_25998.Cx_50.brain_table_merge_output.CPM.NotCtrl.NotEmpty.mat',dir_data),'I_');
load(sprintf('%s/20161026_covariate_table.format.mat',dir_data),'C_');
%%%%%%%%;

%%%%%%%%;
% Here we attempt to extract the gene-lengths. ;
%%%%%%%%;
str_table_name = sprintf('%s/EXON.Gx_50421.Cx_50.brain_table_merge_output.CPM.NotCtrl.NotEmpty.txt',dir_data);
fp = fopen(str_table_name); tmp_str = fgetl(fp); fclose(fp);
E_VariableName_ = strread(tmp_str,'%s','delimiter','\t');
clear tmp_str ;
%%%%%%%%;
str_table_name = sprintf('%s/INTRON_INTERGENIC.Gx_25998.Cx_50.brain_table_merge_output.CPM.NotCtrl.NotEmpty.txt',dir_data);
fp = fopen(str_table_name); tmp_str = fgetl(fp); fclose(fp);
I_VariableName_ = strread(tmp_str,'%s','delimiter','\t');
clear tmp_str ;
%%%%%%%%;
str_table_name = sprintf('%s/ensg_lengths.tsv',dir_data);
fp = fopen(str_table_name); gene_length_ = textscan(fp,'%s %d'); fclose(fp);
for ng=1:length(gene_length_{1});
end;%for ng=1:length(gene_length_{1});
%%%%%%%%;
[tmp_cap_,E_VariableName_to_cap_,gene_length_to_cap_] = intersect(E_VariableName_,gene_length_{1},'stable');
E_length_ = zeros(length(E_VariableName_),1);
E_length_(E_VariableName_to_cap_) = gene_length_{2}(gene_length_to_cap_);
flag_test=1;
if flag_test;
disp(sprintf(' %% testing E_length_ vs gene_length_ '));
n_i=1024;
for ni=1:n_i;
ng = max(1,min(length(tmp_cap_),floor(length(tmp_cap_)*rand())));
assert(strcmp(E_VariableName_(E_VariableName_to_cap_(ng)),gene_length_{1}(gene_length_to_cap_(ng))));
end;%for ni=1:n_i;
end;%if flag_test;
clear tmp_cap_;
%%%%%%%%;
[tmp_cap_,I_VariableName_to_cap_,gene_length_to_cap_] = intersect(I_VariableName_,gene_length_{1},'stable');
I_length_ = zeros(length(I_VariableName_),1);
I_length_(I_VariableName_to_cap_) = gene_length_{2}(gene_length_to_cap_);
flag_test=1;
if flag_test;
disp(sprintf(' %% testing I_length_ vs gene_length_ '));
n_i=1024;
for ni=1:n_i;
ng = max(1,min(length(tmp_cap_),floor(length(tmp_cap_)*rand())));
assert(strcmp(I_VariableName_(I_VariableName_to_cap_(ng)),gene_length_{1}(gene_length_to_cap_(ng))));
end;%for ni=1:n_i;
end;%if flag_test;
clear tmp_cap_;
%%%%%%%%;
% Note that only 6660 (13%) of the E_ are found, and only 5379 (21%) of the I_ are found. ;
%%%%%%%%;

Label_ID_ = unique(C_.Cluster_ID_20161007);
n_Label_ID = length(Label_ID_);
n_Label_ID_ = zeros(n_Label_ID,1);
for nLabel_ID = 1:n_Label_ID;
n_Label_ID_(nLabel_ID) = length(find(strcmp(C_.Cluster_ID_20161007,Label_ID_(nLabel_ID))));
end;%for nLabel_ID = 1:n_Label_ID;
flag_plot=1;
if flag_plot;
bar(1:n_Label_ID,n_Label_ID_);
set(gca,'XTick',1:n_Label_ID,'XTickLabel',Label_ID_); xtickangle(90);
xlabel('cluster label');
ylabel('number');
title('histogram of cluster label counts');
set(gcf,'Position',1+[0,0,1024*2,1024]);
fname_base = sprintf('%s/dir_jpg/label_count_C',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
end;%if flag_plot;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%;
% First pass at defining E_val_ and I_val_ ;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%;
E_ID_ = E_{:,1};
I_ID_ = E_{:,1};
C_ID_ = C_{:,1}; 
u_ID_ = intersectall({E_ID_,I_ID_,C_ID_});
n_u = length(u_ID_);
[~,~,E_to_u_ID_] = intersect(u_ID_,E_ID_,'stable');
[~,~,I_to_u_ID_] = intersect(u_ID_,I_ID_,'stable');
[~,~,C_to_u_ID_] = intersect(u_ID_,C_ID_,'stable');
%%%%%%%%;
flag_test=1;
if flag_test;
disp(sprintf(' %% testing E_to_C_ID_'));
n_i = 1024;
for ni=1:n_i;
nu = max(1,min(n_u,floor(n_u*rand())));
nE_ID = E_to_u_ID_(nu);
nI_ID = I_to_u_ID_(nu);
nC_ID = C_to_u_ID_(nu);
assert(strcmp(C_ID_(nC_ID),E_ID_(nE_ID)));
assert(strcmp(C_ID_(nC_ID),I_ID_(nE_ID)));
end;%for ni=1:n_i;
disp(sprintf(' %% finished testing E_to_C_ID_'));
end;%if flag_test;
%%%%%%%%;
% Extracting unique cluster labels. ;
%%%%%%%%;
u_CLabel_ = unique(C_.Cluster_ID_20161007(C_to_u_ID_));
n_CLabel = length(u_CLabel_);
n_CLabel_ = zeros(n_CLabel,1);
for nCLabel = 1:n_CLabel;
n_CLabel_(nCLabel) = length(find(strcmp(C_.Cluster_ID_20161007(C_to_u_ID_),u_CLabel_(nCLabel))));
end;%for nCLabel = 1:n_CLabel;
flag_plot=1;
if flag_plot;
bar(1:n_CLabel,n_CLabel_);
set(gca,'XTick',1:n_CLabel,'XTickLabel',u_CLabel_); xtickangle(90);
xlabel('cluster label');
ylabel('number');
title('histogram of cluster label counts');
set(gcf,'Position',1+[0,0,1024*2,1024]);
fname_base = sprintf('%s/dir_jpg/label_count_I',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
end;%if flag_plot;
%%%%%%%%;
% Generating ordered matrices of gene-values. ;
%%%%%%%%;
if flag_E;
E_col_val_ = 52:size(E_,2)-1; %<-- gene list begins after 'Fail Confidence Sum' ;
E_val_ = decostand_total_0(E_{E_to_u_ID_,E_col_val_},'col');
n_E_GENE = size(E_val_,2);
end;%if flag_E;
if flag_I;
I_col_val_ = 52:size(I_,2)-1;
I_val_ = decostand_total_0(I_{I_to_u_ID_,I_col_val_},'col');
n_I_GENE = size(I_val_,2);  %<-- gene list begins after 'Fail Confidence Sum' ;
end;%if flag_I;
%assert(size(I_val_,1)==size(E_val_,1)); %<-- fewer samples in E_. ;
%%%%%%%%;
% Now check for bimodality of gene expression across samples: ;
%%%%%%%%;
flag_plot=1;
if flag_plot;
clf;
subplot(2,1,1); tmp_b_ = 0:0.25:6; tmp_h_ = hist(log10(1+E_{E_to_u_ID_,E_col_val_}),tmp_b_); colormap(colormap_beach()); imagesc(log10(1+tmp_h_),[0,3]); xlabel('gene index'); ylabel('log #'); title('E distribution');
subplot(2,1,2); tmp_b_ = 0:0.25:6; tmp_h_ = hist(log10(1+I_{I_to_u_ID_,I_col_val_}),tmp_b_); colormap(colormap_beach()); imagesc(log10(1+tmp_h_),[0,3]); xlabel('gene index'); ylabel('log #'); title('I distribution');
fname_base = sprintf('%s/dir_jpg/expression_histogram',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
end;%if flag_plot;
%%%%%%%%;

%%%%%%%%;
% Determining cutoffs for outliers. ;
%%%%%%%%;
E_val_s1_ = sum(log10(1e-12 + E_val_),1); 
I_val_s1_ = sum(log10(1e-12 + I_val_),1); 
s1_threshold = -1.9e4;
disp(sprintf(' %% s1 threshold %0.2f',s1_threshold));
E_s1_ij_ = find(E_val_s1_>s1_threshold); I_s1_ij_ = find(I_val_s1_>s1_threshold);
%E_s1_ij_ = 1:1e4; I_s1_ij_ = 1:1e4;
E_val_s2_ = sum(log10(1e-12 + E_val_(:,E_s1_ij_)),2);
I_val_s2_ = sum(log10(1e-12 + I_val_(:,I_s1_ij_)),2);
clf;
s2_threshold = so2g_mle_fminsearch(E_val_s2_ + I_val_s2_);
fname_base = sprintf('%s/dir_jpg/expression_s2_histogram',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
disp(sprintf(' %% s2 threshold %0.2f',s2_threshold));
X_s2_ij_ = find(E_val_s2_+I_val_s2_ > s2_threshold);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%;
% Second pass at defining E_val_ and I_val_ ;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%;
u_ID_ = u_ID_(X_s2_ij_);
n_u = length(u_ID_);
[~,~,E_to_u_ID_] = intersect(u_ID_,E_ID_,'stable');
[~,~,I_to_u_ID_] = intersect(u_ID_,I_ID_,'stable');
[~,~,C_to_u_ID_] = intersect(u_ID_,C_ID_,'stable');
%%%%%%%%;
flag_test=1;
if flag_test;
disp(sprintf(' %% testing E_to_C_ID_'));
n_i = 1024;
for ni=1:n_i;
nu = max(1,min(n_u,floor(n_u*rand())));
nE_ID = E_to_u_ID_(nu);
nI_ID = I_to_u_ID_(nu);
nC_ID = C_to_u_ID_(nu);
assert(strcmp(C_ID_(nC_ID),E_ID_(nE_ID)));
assert(strcmp(C_ID_(nC_ID),I_ID_(nE_ID)));
end;%for ni=1:n_i;
disp(sprintf(' %% finished testing E_to_C_ID_'));
end;%if flag_test;
%%%%%%%%;
% Extracting unique cluster labels. ;
%%%%%%%%;
u_CLabel_ = unique(C_.Cluster_ID_20161007(C_to_u_ID_));
n_CLabel = length(u_CLabel_);
n_CLabel_ = zeros(n_CLabel,1);
for nCLabel = 1:n_CLabel;
n_CLabel_(nCLabel) = length(find(strcmp(C_.Cluster_ID_20161007(C_to_u_ID_),u_CLabel_(nCLabel))));
end;%for nCLabel = 1:n_CLabel;
flag_plot=1;
if flag_plot;
bar(1:n_CLabel,n_CLabel_);
set(gca,'XTick',1:n_CLabel,'XTickLabel',u_CLabel_); xtickangle(90);
xlabel('cluster label');
ylabel('number');
title('histogram of cluster label counts');
set(gcf,'Position',1+[0,0,1024*2,1024]);
fname_base = sprintf('%s/dir_jpg/label_count_sub',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
end;%if flag_plot;
%%%%%%%%;
% Generating ordered matrices of gene-values. ;
%%%%%%%%;
if flag_E;
E_col_val_ = 52:size(E_,2)-1; %<-- gene list begins after 'Fail Confidence Sum' ;
E_col_val_ = E_col_val_(E_s1_ij_);
E_val_ = decostand_total_0(E_{E_to_u_ID_,E_col_val_},'col');
n_E_GENE = size(E_val_,2);
end;%if flag_E;
if flag_I;
I_col_val_ = 52:size(I_,2)-1;
I_col_val_ = I_col_val_(I_s1_ij_);
I_val_ = decostand_total_0(I_{I_to_u_ID_,I_col_val_},'col');
n_I_GENE = size(I_val_,2);  %<-- gene list begins after 'Fail Confidence Sum' ;
end;%if flag_I;
assert(size(I_val_,1)==size(E_val_,1)); %<-- same number of samples in I_ and E_. ;
%%%%%%%%;
% Now check for bimodality of gene expression across samples: ;
%%%%%%%%;
flag_plot=1;
if flag_plot;
clf;
subplot(2,1,1); tmp_b_ = 0:0.25:6; tmp_h_ = hist(log10(1+E_{E_to_u_ID_,E_col_val_}),tmp_b_); colormap(colormap_beach()); imagesc(log10(1+tmp_h_),[0,3]); xlabel('gene index'); ylabel('log #'); title('E distribution');
subplot(2,1,2); tmp_b_ = 0:0.25:6; tmp_h_ = hist(log10(1+I_{I_to_u_ID_,I_col_val_}),tmp_b_); colormap(colormap_beach()); imagesc(log10(1+tmp_h_),[0,3]); xlabel('gene index'); ylabel('log #'); title('I distribution');
fname_base = sprintf('%s/dir_jpg/expression_histogram_sub',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
end;%if flag_plot;
%%%%%%%%;

%%%%%%%%;
% Generating rank-normalized matrices. ;
%%%%%%%%;
if flag_E; E_rank_ = rank_normalize_0(E_val_,'row'); end;
if flag_I; I_rank_ = rank_normalize_0(I_val_,'row'); end;

%%%%%%%%;
% Extracting glossary. ;
% The vector 'LC' will list the LinearCorrelationWithPositiveQuality. ;
%%%%%%%%;
str_table_name = sprintf('%s/metric_glossary.txt',dir_data);
G_ = readtable(str_table_name); n_G = size(G_,1);
G_h_ = zeros(255,n_G); 
for nG=1:n_G;
G_h_(:,nG) = hist(cast(cast(G_{nG,1}{1},'uint8'),'double'),1:255);
end;%for nG=1:n_G;
G_h_mask_ = zeros(255,1);
G_h_mask_(cast('0','uint8'):cast('9','uint8')) = 1;
G_h_mask_(cast('A','uint8'):cast('Z','uint8')) = 1;
G_h_mask_(cast('a','uint8'):cast('z','uint8')) = 1;
G_h_ = G_h_.*repmat(G_h_mask_,1,n_G);
for nG=1:n_G;
G_h_(:,nG) = G_h_(:,nG)/norm(G_h_(:,nG),'fro');
end;%for nG=1:n_G;

%%%%%%%%;
% Extracting categorical covariates. ;
%%%%%%%%;
B_col_val_ = setdiff(7:23,[9,10,11,13,15,16,21,22,23]);
% 09 --> cell [1] x3ClassPrediction: 
% 10 <-- FailConfidencSum (double);
% 13 <-- SampleCount (double) ;
% 11 --> cell [47] max_leaf: 
% 15 --> cell [283] well: 
% 16 --> cell [9] batch: 
% 21 --> double [48] SequencingLane: 
% 22 --> double [3] FailState: 
% 23 --> double [12] RunID: 
n_BCOV = length(B_col_val_);
disp(sprintf(' %% %% %% %% '));
for nB=1:n_BCOV;
nc = B_col_val_(nB);
tmp_str = C_.Properties.VariableNames{nc};
tmp_ij = find(strcmp(G_{:,1},tmp_str));
tmp_LC = 0;
if ( isempty(tmp_ij)); disp(sprintf(' %% Warning! %s not found in glossary',tmp_str)); end;
if (~isempty(tmp_ij)); 
tmp_LC_str = G_{tmp_ij,2};
if strcmp(tmp_LC_str,'NA'); tmp_LC = 0; end;
if strcmp(tmp_LC_str,'Unknown'); tmp_LC = 0; end;
if strcmp(tmp_LC_str,'Positive'); tmp_LC = +1; end;
if strcmp(tmp_LC_str,'Negative'); tmp_LC = -1; end;
end;%if (~isempty(tmp_ij));
tmp_type = class(C_{C_to_u_ID_,nc});
tmp_u_ = unique(C_{C_to_u_ID_,nc});
disp(sprintf(' %% %0.2d --> %s [%d] %s [LC %d]: ',nc,tmp_type,length(tmp_u_),tmp_str,tmp_LC));
for nu=1:length(tmp_u_);
if strcmp(tmp_type,'cell');   disp(sprintf(' %% %% %s: %d',tmp_u_{nu},length(find(strcmp(C_{C_to_u_ID_,nc},tmp_u_{nu}))))); end;
if strcmp(tmp_type,'double'); disp(sprintf(' %% %% %d: %d',tmp_u_(nu),length(find(C_{C_to_u_ID_,nc} == tmp_u_(nu))))); end;
end;%for nu=1:length(tmp_u_);
disp(sprintf(' %% %% %% %% '));
end;%for nB=1:n_BCOV;
clear tmp_str tmp_ij tmp_LC tmp_LC_str tmp_type tmp_u_ ;
%%%%%%%%;
% Note that all LC are 0 for the categorical covariates. ;
%%%%%%%%;

%%%%%%%%;
n_FACTOR = 0; B_val_ = zeros(n_u,0); B_VariableName_ = cell(0);
%%%%%%%%;
% Cluster_Grp_20161007:
%%%%%%%%;
nc=7; tmp_str = 'exc'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
nc=7; tmp_str = 'glia'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
nc=7; tmp_str = 'inh'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
%%%%%%%%;
% IsOutlier_20161007: 
%%%%%%%%;
nc=8; tmp_str = 'no'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
nc=8; tmp_str = 'yes'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
%%%%%%%%;
% cell_class: 
%%%%%%%%;
nc=12; tmp_str = 'GABAergic'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
nc=12; tmp_str = 'Glutamatergic'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
nc=12; tmp_str = 'Non-neuronal'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
%%%%%%%%;
% BatchCount: 
%%%%%%%%;
nc=14; tmp_d = 1;
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(C_{C_to_u_ID_,nc}==tmp_d); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
nc=14; tmp_d = 2;
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(C_{C_to_u_ID_,nc}==tmp_d); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
nc=14; tmp_d = 3;
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(C_{C_to_u_ID_,nc}==tmp_d); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
%%%%%%%%;
% neun:
%%%%%%%%;
nc=17; tmp_str = '1NeuNN'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
nc=17; tmp_str = '1NeuNP'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
%%%%%%%%;
% patient_id: 
%%%%%%%%;
nc=18; tmp_str1 = 'H200-1025'; tmp_str2 = 'H200.1025'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str1) | strcmp(C_{C_to_u_ID_,nc},tmp_str2)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
nc=18; tmp_str1 = 'H200-1030'; tmp_str2 = 'H200.1030'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str1) | strcmp(C_{C_to_u_ID_,nc},tmp_str2)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
%%%%%%%%;
% region:
%%%%%%%%;
nc=19; tmp_str = 'FI'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
nc=19; tmp_str = 'MTG'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
%%%%%%%%;
% layer: 
%%%%%%%%;
nc=20; tmp_str = '1'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
nc=20; tmp_str = '5'; 
B_VariableName_{1+n_FACTOR} = sprintf('%s_%s',C_.Properties.VariableNames{nc},tmp_str); tmp_ij_ = find(strcmp(C_{C_to_u_ID_,nc},tmp_str)); B_val_(tmp_ij_,1+n_FACTOR)=1; n_FACTOR = n_FACTOR+1;
%%%%%%%%;
clear tmp_ij_ tmp_str tmp_str1 tmp_str2 ;

%%%%%%%%;
% Ranking categorical covariates. ;
%%%%%%%%;
B_rank_ = rank_normalize_0(B_val_); n_BCOV = size(B_rank_,2);
%%%%%%%%;
% Because LC = 0 for all categorical covariates, we can assign B_LC_ now. ;
%%%%%%%%;
B_LC_ = zeros(n_BCOV,1);

%%%%%%%%;
% Extracting continuous covariates. ;
%%%%%%%%;
D_col_val_ = [10,13,24:size(C_,2)];
str_exclude_ = {'fastQCResult','x_P_E__RawSeq__','x_Num_1_NumberOfInputCoreGenes_CoreGenes_','x_Num_2_NumberOfInputCoreGenes_CoreGenes_','x_MitoCore_NumberOfInputCoreGenes_CoreGenes_','x_MitoCore13_NumberOfInputCoreGenes_CoreGenes_','x_P_E__Pretrimmed_ERCCAligned_'};
for ne=1:length(str_exclude_);
nx = find(strcmp(C_.Properties.VariableNames,str_exclude_{ne}));
disp(sprintf(' %% excluding %s <-- %d <-- %s',str_exclude_{ne},nx,C_.Properties.VariableNames{nx}));
D_col_val_ = setdiff(D_col_val_,nx);
end;%for ne=1:length(str_exclude_);
%%%%%%%%;
n_D_col_val = length(D_col_val_); n_DCOV = n_D_col_val;
D_val_ = zeros(n_u,n_DCOV);
D_LC_ = zeros(n_DCOV,1);
D_VariableName_ = cell(n_D_col_val,1);
for nc=1:n_D_col_val;
D_col_val = D_col_val_(nc);
D_VariableName_{nc} = C_.Properties.VariableNames{D_col_val};
tmp_ij = find(strcmp(G_{:,1},D_VariableName_{nc}));
tmp_LC = 0;
if ( isempty(tmp_ij)); 
disp(sprintf(' %% %s no exact match found in glossary',D_VariableName_{nc})); 
tmp_h_ = hist(cast(cast(D_VariableName_{nc},'uint8'),'double'),1:255);
tmp_h_ = tmp_h_.*transpose(G_h_mask_);
tmp_h_ = tmp_h_/norm(tmp_h_,'fro');
[~,tmp_ij] = max(tmp_h_*G_h_);
disp(sprintf(' %% %s closest match: %s (%d)',D_VariableName_{nc},G_{tmp_ij,1}{1},tmp_ij));
end;%if ( isempty(tmp_ij)); 
if (~isempty(tmp_ij)); 
tmp_LC_str = G_{tmp_ij,2};
if strcmp(tmp_LC_str,'NA'); tmp_LC = 0; end;
if strcmp(tmp_LC_str,'Unknown'); tmp_LC = 0; end;
if strcmp(tmp_LC_str,'Positive'); tmp_LC = +1; end;
if strcmp(tmp_LC_str,'Negative'); tmp_LC = -1; end;
disp(sprintf(' %% %s found in glossary: %d',D_VariableName_{nc},tmp_LC));
end;%if (~isempty(tmp_ij)); 
D_LC_(nc) = tmp_LC;
tmp_D_val_ = C_(C_to_u_ID_,D_col_val); 
if strcmp(class(tmp_D_val_{1,1}),'double'); D_val_(:,nc) = tmp_D_val_{1:end,1};
elseif strcmp(class(tmp_D_val_{1,1}),'cell'); D_val_(:,nc) = cellfun(@str2num,tmp_D_val_{1:end,1}); 
else disp(sprintf(' %% nc %d D_col_val %d LC %d class %s',nc,D_col_val,D_LC_(nc),class(tmp_D_val_{1,1}))); end;
end;%for nc=1:n_D_col_val;
D_rank_ = rank_normalize_0(D_val_);
clear tmp_ij tmp_LC tmp_h_ tmp_LC_str tmp_D_val_ ;

%%%%%%%%;
% Stacking together B_rank_ and D_rank_. ;
%%%%%%%%;
n_CCOV = n_BCOV + n_DCOV;
C_VariableName_ = {B_VariableName_{:},D_VariableName_{:}};
C_val_ = [B_val_ , D_val_];
C_rank_ = [B_rank_ , D_rank_];
C_LC_ = [B_LC_ ; D_LC_];
C_VariableName_LC_ = C_VariableName_;
for nCCOV=1:n_CCOV;
if C_LC_(nCCOV) == -1; tmp_suffix = '[ - ]'; end;
if C_LC_(nCCOV) ==  0; tmp_suffix = '[    ]'; end;
if C_LC_(nCCOV) == +1; tmp_suffix = '[+]'; end;
C_VariableName_LC_{nCCOV} = sprintf('%s_%s',C_VariableName_{nCCOV},tmp_suffix);
end;%for nCCOV=1:n_CCOV;
clear tmp_suffix ;

if flag_E;
%%%%%%%%;
% Now measure correlation across samples of E_ and C_. ;
%%%%%%%%;
tmp_rank_avg_ = mean(C_rank_,1); tmp_rank_std_ = max(1,std(C_rank_,1,1));
tmp_C_ = (C_rank_ - repmat(tmp_rank_avg_,n_u,1))./repmat(tmp_rank_std_,n_u,1);
tmp_rank_avg_ = mean(E_rank_,1); tmp_rank_std_ = max(1,std(E_rank_,1,1));
tmp_E_ = (E_rank_ - repmat(tmp_rank_avg_,n_u,1))./repmat(tmp_rank_std_,n_u,1);
clear tmp_rank_avg_ tmp_rank_std_;
CtEn_rank_ = transpose(tmp_C_)*tmp_E_ / n_u;
[tmp_U_,tmp_S_,tmp_V_] = svds(CtEn_rank_,1); [~,CtEn_rank_U_ij_] = sort(tmp_U_,'descend'); [~,CtEn_rank_V_ij_] = sort(tmp_V_,'descend');
CtEn_rank_ori_ = CtEn_rank_(CtEn_rank_U_ij_,CtEn_rank_V_ij_);
clear CtEn_rank_ tmp_U_ tmp_S_ tmp_V_ ;
flag_disp=1;
if flag_disp;
clf;
colormap(colormap_beach()); 
imagesc(transpose(CtEn_rank_ori_),[-1,+1]); 
set(gca,'XTick',1:n_CCOV,'XTickLabel',C_VariableName_LC_(CtEn_rank_U_ij_),'TickLabelInterpreter','none'); xtickangle(90);
set(gca,'FontSize',7);
xlabel('covariates'); ylabel('genes'); title('Correlation between Genes and Covariates');
set(gcf,'Position',1+[0,0,1024*2,1024]);
colorbar;
fname_base = sprintf('%s/dir_jpg/CtEn_sub_ori',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
end;%if flag_disp;
n_iteration  = 16;
CtEn_rank_avg_ = zeros(n_CCOV,n_E_GENE);
CtEn_rank_std_ = zeros(n_CCOV,n_E_GENE);
for niteration=1:n_iteration;
if (mod(niteration,10)==0); disp(sprintf(' %% niteration %d/%d',niteration,n_iteration)); end;
[tmp_Q_,~] = qr(randn(n_u));
tmp_CtEn_rank_ = transpose(tmp_C_)*tmp_Q_*tmp_E_ / n_u; clear tmp_Q_;
[tmp_U_,tmp_S_,tmp_V_] = svds(tmp_CtEn_rank_,1); [~,tmp_CtEn_rank_U_ij_] = sort(tmp_U_,'descend'); [~,tmp_CtEn_rank_V_ij_] = sort(tmp_V_,'descend');
CtEn_rank_avg_ = CtEn_rank_avg_ + tmp_CtEn_rank_(tmp_CtEn_rank_U_ij_,tmp_CtEn_rank_V_ij_);
CtEn_rank_std_ = CtEn_rank_std_ + tmp_CtEn_rank_(tmp_CtEn_rank_U_ij_,tmp_CtEn_rank_V_ij_).^2;
clear tmp_CtEn_rank_ tmp_U_ tmp_S_ tmp_V_ tmp_CtEn_rank_U_ij_ tmp_CtEn_rank_V_ij_ ;
end;%for niteration=1:n_iteration;
CtEn_rank_avg_ = CtEn_rank_avg_/n_iteration;
CtEn_rank_std_ = sqrt(CtEn_rank_std_/n_iteration - CtEn_rank_avg_.^2);
CtEn_rank_Z_ = ( CtEn_rank_ori_ - CtEn_rank_avg_ ) ./ CtEn_rank_std_ ;
%CtEn_rank_p_ = 0.5 * erfc( CtEn_rank_Z_ / sqrt(2) ) ;
CtEn_rank_p_ = 1.0 * erfc( abs(CtEn_rank_Z_) / sqrt(2) ) ; %<-- two sided. ;
clear tmp_C_ tmp_E_;
flag_disp=1;
if flag_disp;
clf;
colormap(colormap_beach()); 
imagesc(transpose(-log10(CtEn_rank_p_)),[0,+15]); 
set(gca,'XTick',1:n_CCOV,'XTickLabel',C_VariableName_LC_(CtEn_rank_U_ij_),'TickLabelInterpreter','none'); xtickangle(90);
set(gca,'FontSize',7);
xlabel('covariates'); ylabel('genes'); title('log10(p-value) of Correlation between Genes and Covariates');
set(gcf,'Position',1+[0,0,1024*2,1024]);
colorbar;
fname_base = sprintf('%s/dir_jpg/CtEn_sub_p',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
end;%if flag_disp;
end;%if flag_E;

if flag_I;
%%%%%%%%;
% Now measure correlation across samples of I_ and C_. ;
%%%%%%%%;
tmp_rank_avg_ = mean(C_rank_,1); tmp_rank_std_ = max(1,std(C_rank_,1,1));
tmp_C_ = (C_rank_ - repmat(tmp_rank_avg_,n_u,1))./repmat(tmp_rank_std_,n_u,1);
tmp_rank_avg_ = mean(I_rank_,1); tmp_rank_std_ = max(1,std(I_rank_,1,1));
tmp_I_ = (I_rank_ - repmat(tmp_rank_avg_,n_u,1))./repmat(tmp_rank_std_,n_u,1);
clear tmp_rank_avg_ tmp_rank_std_;
CtIn_rank_ = transpose(tmp_C_)*tmp_I_ / n_u;
[tmp_U_,tmp_S_,tmp_V_] = svds(CtIn_rank_,1); [~,CtIn_rank_U_ij_] = sort(tmp_U_,'descend'); [~,CtIn_rank_V_ij_] = sort(tmp_V_,'descend');
CtIn_rank_ori_ = CtIn_rank_(CtIn_rank_U_ij_,CtIn_rank_V_ij_);
clear CtIn_rank_ tmp_U_ tmp_S_ tmp_V_ ;
flag_disp=1;
if flag_disp;
clf;
colormap(colormap_beach()); 
imagesc(transpose(CtIn_rank_ori_),[-1,+1]); 
set(gca,'XTick',1:n_CCOV,'XTickLabel',C_VariableName_LC_(CtIn_rank_U_ij_),'TickLabelInterpreter','none'); xtickangle(90);
set(gca,'FontSize',7);
xlabel('covariates'); ylabel('genes'); title('Correlation between Genes and Covariates');
set(gcf,'Position',1+[0,0,1024*2,1024]);
colorbar;
fname_base = sprintf('%s/dir_jpg/CtIn_sub_ori',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
end;%if flag_disp;
n_iteration  = 16;
CtIn_rank_avg_ = zeros(n_CCOV,n_I_GENE);
CtIn_rank_std_ = zeros(n_CCOV,n_I_GENE);
for niteration=1:n_iteration;
if (mod(niteration,10)==0); disp(sprintf(' %% niteration %d/%d',niteration,n_iteration)); end;
[tmp_Q_,~] = qr(randn(n_u));
tmp_CtIn_rank_ = transpose(tmp_C_)*tmp_Q_*tmp_I_ / n_u; clear tmp_Q_;
[tmp_U_,tmp_S_,tmp_V_] = svds(tmp_CtIn_rank_,1); [~,tmp_CtIn_rank_U_ij_] = sort(tmp_U_,'descend'); [~,tmp_CtIn_rank_V_ij_] = sort(tmp_V_,'descend');
CtIn_rank_avg_ = CtIn_rank_avg_ + tmp_CtIn_rank_(tmp_CtIn_rank_U_ij_,tmp_CtIn_rank_V_ij_);
CtIn_rank_std_ = CtIn_rank_std_ + tmp_CtIn_rank_(tmp_CtIn_rank_U_ij_,tmp_CtIn_rank_V_ij_).^2;
clear tmp_CtIn_rank_ tmp_U_ tmp_S_ tmp_V_ tmp_CtIn_rank_U_ij_ tmp_CtIn_rank_V_ij_ ;
end;%for niteration=1:n_iteration;
CtIn_rank_avg_ = CtIn_rank_avg_/n_iteration;
CtIn_rank_std_ = sqrt(CtIn_rank_std_/n_iteration - CtIn_rank_avg_.^2);
CtIn_rank_Z_ = ( CtIn_rank_ori_ - CtIn_rank_avg_ ) ./ CtIn_rank_std_ ;
%CtIn_rank_p_ = 0.5 * erfc( CtIn_rank_Z_ / sqrt(2) ) ;
CtIn_rank_p_ = 1.0 * erfc( abs(CtIn_rank_Z_) / sqrt(2) ) ; %<-- two sided. ;
clear tmp_C_ tmp_I_;
flag_disp=1;
if flag_disp;
clf;
colormap(colormap_beach()); 
imagesc(transpose(-log10(CtIn_rank_p_)),[0,+15]); 
set(gca,'XTick',1:n_CCOV,'XTickLabel',C_VariableName_LC_(CtIn_rank_U_ij_),'TickLabelInterpreter','none'); xtickangle(90);
set(gca,'FontSize',7);
xlabel('covariates'); ylabel('genes'); title('log10(p-value) of Correlation between Genes and Covariates');
set(gcf,'Position',1+[0,0,1024*2,1024]);
colorbar;
fname_base = sprintf('%s/dir_jpg/CtIn_sub_p',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
end;%if flag_disp;
end;%if flag_I;

flag_disp=1;
if flag_disp;
clf;
n_bin = 32;
%%%%%%%%;
if flag_E;
subplot(2,4,1); hold on;
tmp_ng_ = 1:n_E_GENE;
%plot(tmp_ng_,log10(1+E_length_(E_col_val_(CtEn_rank_V_ij_))),'r.');
tmp_ij_ = find(E_length_(E_col_val_(CtEn_rank_V_ij_))>0);
plot(tmp_ng_(tmp_ij_),log10(1+E_length_(E_col_val_(CtEn_rank_V_ij_(tmp_ij_)))),'ro');
xlabel('gene-index (matching CtEn_p order)'); xlim([0,n_E_GENE]); tmp_yl_ = ylim();
ylabel('log10(gene length)');
title('E gene length vs gene ordering');
subplot(2,4,2); hold on;
tmp_ng_ = 1:n_E_GENE;
tmp_ij_ = find(E_length_(E_col_val_(CtEn_rank_V_ij_))>0);
colormap(colormap_beach());
imagesc(log10(1+hist2d_0(tmp_ng_(tmp_ij_),log10(1+E_length_(E_col_val_(CtEn_rank_V_ij_(tmp_ij_)))),n_bin,n_bin,[1,n_E_GENE],tmp_yl_)));
colorbar();
xlabel('gene-index (matching CtEn_p order)'); ylabel('log10(gene length)');
xlim([1,n_bin]); ylim([1,n_bin]);
set(gca,'XTick',[],'XTickLabel',[],'YTick',[],'YTickLabel',[]);
title('E gene length vs gene ordering');
end;%if flag_E;
%%%%%%%%;
if flag_I;
subplot(2,4,3); hold on;
tmp_ng_ = 1:n_I_GENE;
%plot(tmp_ng_,log10(1+I_length_(I_col_val_(CtIn_rank_V_ij_))),'b.');
tmp_ij_ = find(I_length_(I_col_val_(CtIn_rank_V_ij_))>0);
plot(tmp_ng_(tmp_ij_),log10(1+I_length_(I_col_val_(CtIn_rank_V_ij_(tmp_ij_)))),'bo');
xlabel('gene-index (matching CtIn_p order)'); xlim([0,n_I_GENE]); tmp_yl_ = ylim();
ylabel('log10(gene length)');
title('I gene length vs gene ordering');
subplot(2,4,4); hold on;
tmp_ng_ = 1:n_I_GENE;
tmp_ij_ = find(I_length_(I_col_val_(CtIn_rank_V_ij_))>0);
colormap(colormap_beach()); 
imagesc(log10(1+hist2d_0(tmp_ng_(tmp_ij_),log10(1+I_length_(I_col_val_(CtIn_rank_V_ij_(tmp_ij_)))),n_bin,n_bin,[1,n_I_GENE],tmp_yl_)));
colorbar();
xlabel('gene-index (matching CtIn_p order)'); ylabel('log10(gene length)');
xlim([1,n_bin]); ylim([1,n_bin]);
set(gca,'XTick',[],'XTickLabel',[],'YTick',[],'YTickLabel',[]);
title('I gene length vs gene ordering');
end;%if flag_E;
%%%%%%%%;
if flag_E;
subplot(2,4,5); hold on;
tmp_ng_ = 1:n_E_GENE;
tmp_ij_ = find(E_length_(E_col_val_(CtEn_rank_V_ij_))>0);
tmp_X_ = corr(transpose(log10(sum(abs(CtEn_rank_ori_(:,tmp_ij_)).^2,1))),log10(E_length_(E_col_val_(CtEn_rank_V_ij_(tmp_ij_)))));
plot(log10(sum(abs(CtEn_rank_ori_(:,tmp_ij_)).^2,1)),log10(E_length_(E_col_val_(CtEn_rank_V_ij_(tmp_ij_)))),'ro');
xlabel('log10(sum squared correlation)'); ylabel('log10(gene length)');
tmp_yl_ = ylim(); tmp_xl_ = xlim();
title(sprintf('avg-corr^2 vs E gene-length (C %0.6f)',tmp_X_));
subplot(2,4,6); hold on;
colormap(colormap_beach());
imagesc(log10(1+hist2d_0(log10(sum(abs(CtEn_rank_ori_(:,tmp_ij_)).^2,1)),log10(E_length_(E_col_val_(CtEn_rank_V_ij_(tmp_ij_)))),n_bin,n_bin,tmp_xl_,tmp_yl_)));
colorbar();
colorbar();
xlabel('log10(sum squared correlation)'); ylabel('log10(gene length)');
xlim([1,n_bin]); ylim([1,n_bin]);
set(gca,'XTick',[],'XTickLabel',[],'YTick',[],'YTickLabel',[]);
title(sprintf('avg-corr^2 vs E gene-length (C %0.6f)',tmp_X_));
end;%if flag_E;
%%%%%%%%;
if flag_I;
subplot(2,4,7); hold on;
tmp_ng_ = 1:n_I_GENE;
tmp_ij_ = find(I_length_(I_col_val_(CtIn_rank_V_ij_))>0);
tmp_X_ = corr(transpose(log10(sum(abs(CtIn_rank_ori_(:,tmp_ij_)).^2,1))),log10(I_length_(I_col_val_(CtIn_rank_V_ij_(tmp_ij_)))));
plot(log10(sum(abs(CtIn_rank_ori_(:,tmp_ij_)).^2,1)),log10(I_length_(I_col_val_(CtIn_rank_V_ij_(tmp_ij_)))),'bo');
xlabel('log10(sum squared correlation)'); ylabel('log10(gene length)');
tmp_yl_ = ylim(); tmp_xl_ = xlim();
title(sprintf('avg-corr^2 vs I gene-length (C %0.6f)',tmp_X_));
subplot(2,4,8); hold on;
colormap(colormap_beach());
imagesc(log10(1+hist2d_0(log10(sum(abs(CtIn_rank_ori_(:,tmp_ij_)).^2,1)),log10(I_length_(I_col_val_(CtIn_rank_V_ij_(tmp_ij_)))),n_bin,n_bin,tmp_xl_,tmp_yl_)));
colorbar();
colorbar();
xlabel('log10(sum squared correlation)'); ylabel('log10(gene length)');
xlim([1,n_bin]); ylim([1,n_bin]);
set(gca,'XTick',[],'XTickLabel',[],'YTick',[],'YTickLabel',[]);
title(sprintf('avg-corr^2 vs I gene-length (C %0.6f)',tmp_X_));
end;%if flag_I;
%%%%%%%%;
set(gcf,'Position',1+[0,0,1024*2,1024]);
fname_base = sprintf('%s/dir_jpg/CtXn_X_sub_vs_length',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
clear tmp_xl_ tmp_yl_ tmp_X_ tmp_ij_ tmp_ng_ ;
end;%if flag_disp;
%%%%%%%%;
% Note that there is not much correlation between length and covariate-correlation. ;
%%%%%%%%;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%;
% Fit linear model ;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%;

%%%%%%%%;
% Here we try and estimate the genes E_ or I_ using the covariates C_. ;
% For now we use all of C_ to estimate E_ and I_. ;
% Eventually, we might want to use only the C_ which are not associated with quality. ;
%%%%%%%%;
n_rank = 3+3;
%%%%%%%%;
% Can we estimate the link E_rank_ = C_rank_ * zeta_E_  using a low-rank zeta_E_ ?
%%%%%%%%;
if flag_E;
zeta_E_un_ = zeros(1+n_CCOV,n_rank);
zeta_E_vn_ = zeros(n_E_GENE,n_rank);
tmp_Bn_ = pinv([ones(n_u,1) , C_rank_]);
tmp_Yn_ = E_rank_;
for nrank=1:n_rank;
[tmp_un_,tmp_vn_] = rrr_0([ones(n_u,1) , C_rank_],tmp_Yn_,tmp_Bn_);
zeta_E_un_(:,nrank) = tmp_un_;
zeta_E_vn_(:,nrank) = tmp_vn_;
tmp_Yn_ = tmp_Yn_ - (ones(n_u,1)*zeta_E_un_(1,nrank) + C_rank_*zeta_E_un_(2:end,nrank))*transpose(zeta_E_vn_(:,nrank));
clear tmp_un_ tmp_vn_;
end;%for nrank=1:n_rank;
clear tmp_Yn_ tmp_Bn_;
end;%if flag_E;
%%%%%%%%
% Can we estimate the link I_rank_ = C_rank_ * zeta_I_ using a low-rank zeta_I_ ?
%%%%%%%%
if flag_I;
zeta_I_un_ = zeros(1+n_CCOV,n_rank);
zeta_I_vn_ = zeros(n_I_GENE,n_rank);
tmp_Bn_ = pinv([ones(n_u,1) , C_rank_]);
tmp_Yn_ = I_rank_;
for nrank=1:n_rank;
[tmp_un_,tmp_vn_] = rrr_0([ones(n_u,1) , C_rank_],tmp_Yn_,tmp_Bn_);
zeta_I_un_(:,nrank) = tmp_un_;
zeta_I_vn_(:,nrank) = tmp_vn_;
tmp_Yn_ = tmp_Yn_ - (ones(n_u,1)*zeta_I_un_(1,nrank) + C_rank_*zeta_I_un_(2:end,nrank))*transpose(zeta_I_vn_(:,nrank));
clear tmp_un_ tmp_vn_;
end;%for nrank=1:n_rank;
clear tmp_Yn_ tmp_Bn_;
end;%if flag_I;
%%%%%%%%;
% Appears to be rank 3 or so ;
%%%%%%%%;
flag_disp=1;
if flag_disp;
if flag_E;
subplot(2,2,1);
bar(1:n_rank,sqrt(sum((ones(n_u,1)*zeta_E_un_(1,:) + C_rank_*zeta_E_un_(2:end,:)).^2,1)),'r');
xlabel('pc #'); ylabel('l2-norm');
title('numerical rank of zeta_E_','Interpreter','none');
subplot(2,2,3);
bar(1:n_rank,log10(sqrt(sum((ones(n_u,1)*zeta_E_un_(1,:) + C_rank_*zeta_E_un_(2:end,:)).^2,1))),'r');
xlabel('pc #'); ylabel('log10(l2-norm)');
title('numerical rank of zeta_E_','Interpreter','none');
end;%if flag_E;
if flag_I;
subplot(2,2,2);
bar(1:n_rank,sqrt(sum((ones(n_u,1)*zeta_I_un_(1,:) + C_rank_*zeta_I_un_(2:end,:)).^2,1)),'b');
xlabel('pc #'); ylabel('l2-norm');
title('numerical rank of zeta_I_','Interpreter','none');
subplot(2,2,4);
bar(1:n_rank,log10(sqrt(sum((ones(n_u,1)*zeta_I_un_(1,:) + C_rank_*zeta_I_un_(2:end,:)).^2,1))),'b');
xlabel('pc #'); ylabel('log10(l2-norm)');
title('numerical rank of zeta_I_','Interpreter','none');
end;%if flag_I;
set(gcf,'Position',1+[0,0,512*2,512]);
fname_base = sprintf('%s/dir_jpg/zeta_X_sub_',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
end;%if flag_disp;
%%%%%%%%;
n_zeta_E = 3;
n_zeta_I = 3;

flag_disp=1;
if flag_disp;
clf;
if flag_E;
subplot(1,2,1); tmp_ = (ones(n_u,1)*zeta_E_un_(1,:) + C_rank_*zeta_E_un_(2:end,:))*transpose(zeta_E_vn_); hist(tmp_(:),32);
tmp_n = length(find(tmp_(:)<1 | tmp_(:)>n_E_GENE));
disp(sprintf(' %% zeta_E_: %d/%d = %0.2f out of bounds',tmp_n,length(tmp_(:)),tmp_n/length(tmp_(:))))
end;%if flag_E;
if flag_I;
subplot(1,2,2); tmp_ = (ones(n_u,1)*zeta_I_un_(1,:) + C_rank_*zeta_I_un_(2:end,:))*transpose(zeta_I_vn_); hist(tmp_(:),32);
tmp_n = length(find(tmp_(:)<1 | tmp_(:)>n_I_GENE));
disp(sprintf(' %% zeta_I_: %d/%d = %0.2f out of bounds',tmp_n,length(tmp_(:)),tmp_n/length(tmp_(:))))
end;%if flag_I;
clear tmp_ tmp_n;
end;%if flag_disp;
%%%%%%%%;
% only 0.00 out of bounds. ;
%%%%%%%%;

%%%%%%%%;
% Here we try and estimate the covariates C_ using the genes E_ or I_. ;
% For now we estimate all of C_. ;
% Eventually, we might want to only estimate the C_ which are not associated with a quality. ;
%%%%%%%%;
n_rank = 6+3;
%%%%%%%%;
% Can we estimate the link E_rank_ * beta_E_ = C_rank_ using a low-rank beta_E_ ?
%%%%%%%%;
if flag_E;
beta_E_un_ = zeros(1+n_E_GENE,n_rank);
beta_E_vn_ = zeros(n_CCOV,n_rank);
tmp_Bn_ = pinv([ones(n_u,1) , E_rank_]);
tmp_Yn_ = C_rank_;
for nrank=1:n_rank;
[tmp_un_,tmp_vn_] = rrr_0([ones(n_u,1) , E_rank_],tmp_Yn_,tmp_Bn_);
beta_E_un_(:,nrank) = tmp_un_;
beta_E_vn_(:,nrank) = tmp_vn_;
tmp_Yn_ = tmp_Yn_ - (ones(n_u,1)*beta_E_un_(1,nrank) + E_rank_*beta_E_un_(2:end,nrank))*transpose(beta_E_vn_(:,nrank));
clear tmp_un_ tmp_vn_;
end;%for nrank=1:n_rank;
clear tmp_Yn_ tmp_Bn_;
end;%if flag_E;
%%%%%%%%
% Can we estimate the link I_rank_ * beta_I_ = C_rank_ using a low-rank beta_I_ ?
%%%%%%%%
if flag_I;
beta_I_un_ = zeros(1+n_I_GENE,n_rank);
beta_I_vn_ = zeros(n_CCOV,n_rank);
tmp_Bn_ = pinv([ones(n_u,1) , I_rank_]);
tmp_Yn_ = C_rank_;
for nrank=1:n_rank;
[tmp_un_,tmp_vn_] = rrr_0([ones(n_u,1) , I_rank_],tmp_Yn_,tmp_Bn_);
beta_I_un_(:,nrank) = tmp_un_;
beta_I_vn_(:,nrank) = tmp_vn_;
tmp_Yn_ = tmp_Yn_ - (ones(n_u,1)*beta_I_un_(1,nrank) + I_rank_*beta_I_un_(2:end,nrank))*transpose(beta_I_vn_(:,nrank));
clear tmp_un_ tmp_vn_;
end;%for nrank=1:n_rank;
clear tmp_Yn_ tmp_Bn_;
end;%if flag_I;
%%%%%%%%;
% Appears to be rank 6 or so ;
%%%%%%%%;
flag_disp=1;
if flag_disp;
if flag_E;
subplot(2,2,1);
bar(1:n_rank,sqrt(sum((ones(n_u,1)*beta_E_un_(1,:) + E_rank_*beta_E_un_(2:end,:)).^2,1)),'r');
xlabel('pc #'); ylabel('l2-norm');
title('numerical rank of beta_E_','Interpreter','none');
subplot(2,2,3);
bar(1:n_rank,log10(sqrt(sum((ones(n_u,1)*beta_E_un_(1,:) + E_rank_*beta_E_un_(2:end,:)).^2,1))),'r');
xlabel('pc #'); ylabel('log10(l2-norm)');
title('numerical rank of beta_E_','Interpreter','none');
end;%if flag_E;
if flag_I;
subplot(2,2,2);
bar(1:n_rank,sqrt(sum((ones(n_u,1)*beta_I_un_(1,:) + I_rank_*beta_I_un_(2:end,:)).^2,1)),'b');
xlabel('pc #'); ylabel('l2-norm');
title('numerical rank of beta_I_','Interpreter','none');
subplot(2,2,4);
bar(1:n_rank,log10(sqrt(sum((ones(n_u,1)*beta_I_un_(1,:) + I_rank_*beta_I_un_(2:end,:)).^2,1))),'b');
xlabel('pc #'); ylabel('log10(l2-norm)');
title('numerical rank of beta_I_','Interpreter','none');
end;%if flag_I;
set(gcf,'Position',1+[0,0,512*2,512]);
fname_base = sprintf('%s/dir_jpg/beta_X_sub_',dir_trunk);
disp(sprintf(' %% writing %s.jpg',fname_base));
print('-djpeg',sprintf('%s.jpg',fname_base));
print('-depsc',sprintf('%s.eps',fname_base));
end;%if flag_disp;
%%%%%%%%;
n_beta_E = 6;
n_beta_I = 6;

flag_disp=0;
if flag_disp;
clf;
subplot(2,2,1); imagesc(C_rank_); title('Cn');
subplot(2,2,2); imagesc((ones(n_u,1)*beta_I_un_(1,:) + I_rank_*beta_I_un_(2:end,:))*transpose(beta_I_vn_)); title('In*betan');
subplot(2,2,3); imagesc(transpose(I_rank_)*C_rank_); title('It*Cn');
subplot(2,2,4); imagesc(transpose(I_rank_)*((ones(n_u,1)*beta_I_un_(1,:) + I_rank_*beta_I_un_(2:end,:))*transpose(beta_I_vn_))); title('It*In*betan');
end;%if flag_disp;
%%%%%%%%;
% is a logistic model necessary, or does linear model fit well? 
%%%%%%%%;
flag_disp=1;
if flag_disp;
clf;
if flag_E;
subplot(1,2,1); tmp_ = (ones(n_u,1)*beta_E_un_(1,:) + E_rank_*beta_E_un_(2:end,:))*transpose(beta_E_vn_); hist(tmp_(:),32);
tmp_n = length(find(tmp_(:)<1 | tmp_(:)>n_u));
disp(sprintf(' %% beta_E_: %d/%d = %0.2f out of bounds',tmp_n,length(tmp_(:)),tmp_n/length(tmp_(:))))
end;%if flag_E;
if flag_I;
subplot(1,2,2); tmp_ = (ones(n_u,1)*beta_I_un_(1,:) + I_rank_*beta_I_un_(2:end,:))*transpose(beta_I_vn_); hist(tmp_(:),32);
tmp_n = length(find(tmp_(:)<1 | tmp_(:)>n_u));
disp(sprintf(' %% beta_I_: %d/%d = %0.2f out of bounds',tmp_n,length(tmp_(:)),tmp_n/length(tmp_(:))))
end;%if flag_I;
clear tmp_ tmp_n;
end;%if flag_disp;
%%%%%%%%;
% only 0.03 out of bounds. ;
%%%%%%%%;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%;
% Now performing clustering on E_rank_ and I_rank_ after appropriate normalization. ;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%;

E_array_ = E_rank_; %<-- copy. ;
E_array_ = E_array_ - [ (ones(n_u,1)*zeta_E_un_(1,:) + C_rank_*zeta_E_un_(2:end,:))*transpose(zeta_E_vn_) ]; %<-- correct for linear effect of covariates. ;
E_array_ = E_array_ - median(E_array_,1); %<-- normalize for uniform sparsity = 0.5. ;
%E_array_ = E_array_ - mean(E_array_,1); tmp_ = sum(E_array_>0); [~,tmp_ij_] = sort(tmp_,'ascend'); E_array_ = E_array_(:,tmp_ij_); %<-- normalize for variable sparsity (not recommended). ;
%E_array_ = 2*(E_array_>0)-1; %<-- Binarization now is unnecessary, since binarization is performed later on. ;
%n_tmp = size(E_array_,2); E_array_ = E_array_*sparse(1:n_tmp,1:n_tmp,1./sqrt(mean(E_array_.^2,1))); %<-- can magnify certain values tremendously (not recommended) but will not affect data post-binarization. ;
%%%%%%%%;
dir_code = '/data/rangan/dir_bcc/dir_lakcluster_c_dev'; dir_trunk = pwd;
rev_flag = 0;
A_n_rind_ = {[1:n_u]};
A_n_cind = 1:n_E_GENE;
Z_n_rind_ = {[]};
T_n_ = {ones(n_u,1)};
T_n_cind = 1;
GLOBAL_TEST_sparse = 1;
gamma = 0.01;
B_MLT = 34;
Ireq = 0;
shuffle_num = 0;
verbose_flag = 0;
force_create_flag = 1;
prefix = 'test_lakcluster';
lakcluster_uADZSZDA_dr_3(dir_code,dir_trunk,prefix,{E_array_},rev_flag,A_n_rind_,A_n_cind,Z_n_rind_,T_n_,T_n_cind,GLOBAL_TEST_sparse,gamma,B_MLT,Ireq,shuffle_num,verbose_flag,force_create_flag);
out_xdrop_E_0_ = textread('/data/rangan/dir_bcc/dir_jamison/dir_test_lakcluster/dir_test_lakcluster_lak_uADZSZDA_D_g010_B34_I0_s0000/out_xdrop_a.txt');
rdrop_E_0_ = 1+out_xdrop_E_0_(find(out_xdrop_E_0_(:,1)>-1),1); rkeep_E_0_ = rdrop_E_0_(end:-1:1);
cdrop_E_0_ = 1+out_xdrop_E_0_(find(out_xdrop_E_0_(:,2)>-1),2); ckeep_E_0_ = cdrop_E_0_(end:-1:1);
figure;clf;colormap(colormap_beach()); imagesc(E_array_(rkeep_E_0_,ckeep_E_0_)); colorbar; title('lakcluster');
n_i = 16;
for ni=1:n_i;
tmp_prefix = sprintf('%s_s%0.4d',prefix,ni);
disp(sprintf(' %% ni %d/%d %s',ni,n_i,tmp_prefix));
rng(ni); [~,tmp_Q_] = qr(randn(n_u));
tmp_E_array_ = E_rank_; %<-- copy. ;
tmp_E_array_ = tmp_E_array_ - [ (ones(n_u,1)*zeta_E_un_(1,:) + C_rank_*zeta_E_un_(2:end,:))*transpose(zeta_E_vn_) ]; %<-- correct for linear effect of covariates. ;
tmp_E_array_ = tmp_Q_*tmp_E_array_;
tmp_E_array_ = tmp_E_array_ - median(tmp_E_array_,1); %<-- normalize
lakcluster_uADZSZDA_dr_3(dir_code,dir_trunk,tmp_prefix,{tmp_E_array_},rev_flag,A_n_rind_,A_n_cind,Z_n_rind_,T_n_,T_n_cind,GLOBAL_TEST_sparse,gamma,B_MLT,Ireq,shuffle_num,verbose_flag,force_create_flag);
end;%for ni=1:n_i;
%%%%%%%%;
ni=0;
trace_{1+ni} = textread('/data/rangan/dir_bcc/dir_jamison/dir_test_lakcluster/dir_test_lakcluster_lak_uADZSZDA_D_g010_B34_I0_s0000/out_trace.txt');
for ni=1:n_i;
trace_{1+ni} = textread(sprintf('/data/rangan/dir_bcc/dir_jamison/dir_test_lakcluster_s%.4d/dir_test_lakcluster_s%.4d_lak_uADZSZDA_D_g010_B34_I0_s0000/out_trace.txt',ni,ni));
end;%for ni=1:n_i;

hold on;
ni=0;
plot(trace_{1+ni}(:,4),'r-');
for ni=1:n_i;
plot(trace_{1+ni}(:,4),'k-');
end;%for ni=1:n_i;
hold off;

I_array_ = I_rank_; %<-- copy. ;
I_array_ = I_array_ - [ (ones(n_u,1)*zeta_I_un_(1,:) + C_rank_*zeta_I_un_(2:end,:))*transpose(zeta_I_vn_) ]; %<-- correct for linear effect of covariates. ;
%I_array_ = I_array_(:,1:1000); %<-- throw away many genes. ;
I_array_ = I_array_ - median(I_array_,1); %<-- normalize
%I_array_ = 2*(I_array_>0)-1; %<-- Binarize? ;
%n_tmp = size(I_array_,2); I_array_ = I_array_*sparse(1:n_tmp,1:n_tmp,1./sqrt(mean(I_array_.^2,1))); %<-- can magnify certain values tremendously. ;
%%%%%%%%;

E_array_ = E_rank_; %<-- copy. ;
E_array_ = E_array_ - [ (ones(n_u,1)*zeta_E_un_(1,:) + C_rank_*zeta_E_un_(2:end,:))*transpose(zeta_E_vn_) ]; %<-- correct for linear effect of covariates. ;
%E_array_ = E_array_(:,1:1000); %<-- throw away many genes. ;
E_array_ = E_array_ - mean(E_array_,1); %<-- normalize
%E_array_ = 2*(E_array_>0)-1; %<-- Binarize? ;
%n_tmp = size(E_array_,2); E_array_ = E_array_*sparse(1:n_tmp,1:n_tmp,1./sqrt(mean(E_array_.^2,1))); %<-- can magnify certain values tremendously. ;
I_array_ = I_rank_; %<-- copy. ;
I_array_ = I_array_ - [ (ones(n_u,1)*zeta_I_un_(1,:) + C_rank_*zeta_I_un_(2:end,:))*transpose(zeta_I_vn_) ]; %<-- correct for linear effect of covariates. ;
%I_array_ = I_array_(:,1:1000); %<-- throw away many genes. ;
I_array_ = I_array_ - mean(I_array_,1); %<-- normalize
%I_array_ = 2*(I_array_>0)-1; %<-- Binarize? ;
%n_tmp = size(I_array_,2); I_array_ = I_array_*sparse(1:n_tmp,1:n_tmp,1./sqrt(mean(I_array_.^2,1))); %<-- can magnify certain values tremendously. ;
X_array_ = [ E_array_ , I_array_ ];
%%%%%%%%;


disp('returning'); return;



fname_readme = sprintf('%s/dir_mat/readme.txt',dir_trunk);
fp = fopen(fname_readme,'w');
fprintf(fp,'u_ID_.nsv: newline delimited list of sample ids.\n');
fprintf(fp,'C_VariableName_.tsv: tab delimited list of metadata (i.e., Covariates) names.\n');
fprintf(fp,'C_VariableName_.nsv: newline delimited list of metadata (i.e., Covariates) names.\n');
fprintf(fp,'E_GeneName_.nsv: newline delimited list of exon gene names.\n');
fprintf(fp,'I_GeneName_.nsv: newline delimited list of intron gene names.\n');
fprintf(fp,'C_rank_.tsv: tab delimited array of samples (rows) by covariate ranks (columns).\n');
fprintf(fp,'E_rank_.tsv: tab delimited array of samples (rows) by exon gene ranks (columns).\n');
fprintf(fp,'I_rank_.tsv: tab delimited array of smaples (rows) by intron gene ranks (columns).\n');
fprintf(fp,'zeta_E_un_.tsv and zeta_E_vn_.tsv: tab delimited arrays of coefficients. The outer-product [zeta_E_] := [zeta_E_un_]*[zeta_E_vn_] is a coefficient matrix for the linear model [Exon-Gene-Ranks] = [1 , Covariate-Ranks]*[zeta_E_]. The ''1'' in the previous expression refers to a vector of all ones (i.e., the constant term in the linear model).\n');
fprintf(fp,'beta_E_un_.tsv and beta_E_vn_.tsv: tab delimited arrays of coefficients. The outer-product [beta_E_] := [beta_E_un_]*[beta_E_vn_] is a coefficient matrix for the linear model [Covariate-Ranks] = [1 , Exon-Gene-Ranks]*[beta_E_]. The ''1'' in the previous expression refers to a vector of all ones (i.e., the constant term in the linear model).\n');
fprintf(fp,'zeta_I_un_.tsv and zeta_I_vn_.tsv: tab delimited arrays of coefficients. The outer-product [zeta_I_] := [zeta_I_un_]*[zeta_I_vn_] is a coefficient matrix for the linear model [Intron-Gene-Ranks] = [1 , Covariate-Ranks]*[zeta_I_]. The ''1'' in the previous expression refers to a vector of all ones (i.e., the constant term in the linear model).\n');
fprintf(fp,'beta_I_un_.tsv and beta_I_vn_.tsv: tab delimited arrays of coefficients. The outer-product [beta_I_] := [beta_I_un_]*[beta_I_vn_] is a coefficient matrix for the linear model [Covariate-Ranks] = [1 , Intron-Gene-Ranks]*[beta_I_]. The ''1'' in the previous expression refers to a vector of all ones (i.e., the constant term in the linear model).\n');
fprintf(fp,'E_CLabel_pair_.nsv: newline delimited list of cluster-pairs for use with E_AB_C_absZ_.tsv, etc.\n');
fprintf(fp,'I_CLabel_pair_.nsv: newline delimited list of cluster-pairs for use with I_AB_C_absZ_.tsv, etc.. Should be identical to E_CLabel_pair_.nsv.\n');
fprintf(fp,'E_AB_C_absZ_.tsv: tab delimited array of cluster-pairs (rows) by covariate (columns). Each array entry indicates the absolute-value of the z-score associated with the (raw) differential-expression of that particular covariate-rank over that cluster pair.\n');
fprintf(fp,'E_AB_K_absZ_.tsv: tab delimited array of cluster-pairs (rows) by covariate (columns). Each array entry indicates the absolute-value of the z-score associated with the (relative) differential-expression of that particular covariate-rank over that cluster pair. In this context ''relative'' means corrected for the linear effect of the E gene-ranks.\n');
fprintf(fp,'E_AB_Z_absZ_.tsv: tab delimited array of cluster-pairs (rows) by exon genes (columns). Each array entry indicates the absolute-value of the z-score associated with the (raw) differential-expression of that particular gene-rank over that cluster pair.\n');
fprintf(fp,'E_AB_H_absZ_.tsv: tab delimited array of cluster-pairs (rows) by exon genes (columns). Each array entry indicates the absolute-value of the z-score associated with the (relative) differential-expression of that particular gene-rank over that cluster pair. In this context ''relative'' means corrected for the linear effect of the covariate-rank.\n');
fprintf(fp,'I_AB_C_absZ_.tsv: tab delimited array of cluster-pairs (rows) by covariate (columns). Each array entry indicates the absolute-value of the z-score associated with the (raw) differential-expression of that particular covariate-rank over that cluster pair.\n');
fprintf(fp,'I_AB_K_absZ_.tsv: tab delimited array of cluster-pairs (rows) by covariate (columns). Each array entry indicates the absolute-value of the z-score associated with the (relative) differential-expression of that particular covariate-rank over that cluster pair. In this context ''relative'' means corrected for the linear effect of the I gene-ranks.\n');
fprintf(fp,'I_AB_Z_absZ_.tsv: tab delimited array of cluster-pairs (rows) by intron genes (columns). Each array entry indicates the absolute-value of the z-score associated with the (raw) differential-expression of that particular gene-rank over that cluster pair.\n');
fprintf(fp,'I_AB_H_absZ_.tsv: tab delimited array of cluster-pairs (rows) by intron genes (columns). Each array entry indicates the absolute-value of the z-score associated with the (relative) differential-expression of that particular gene-rank over that cluster pair. In this context ''relative'' means corrected for the linear effect of the covariate-rank.\n');
fclose(fp);


%%%%%%%%;
fname_tsv = sprintf('%s/dir_mat/u_ID_.nsv',dir_trunk);
fp=fopen(fname_tsv,'w');
fprintf(fp,'%s\n',u_ID_{1:end});
fclose(fp);
fname_tsv = sprintf('%s/dir_mat/C_VariableName_.tsv',dir_trunk);
fp=fopen(fname_tsv,'w');
fprintf(fp,'%s\t',C_VariableName_{1:end});
fclose(fp);
fname_tsv = sprintf('%s/dir_mat/C_VariableName_.nsv',dir_trunk);
fp=fopen(fname_tsv,'w');
fprintf(fp,'%s\n',C_VariableName_{1:end});
fclose(fp);
%%%%%%%%;
fname_tsv = sprintf('%s/dir_mat/E_GeneName_.nsv',dir_trunk);
fp=fopen(fname_tsv,'w');
fprintf(fp,'%s\n',E_VariableName_{52:end});
fclose(fp);
%%%%%%%%;
fname_tsv = sprintf('%s/dir_mat/I_GeneName_.nsv',dir_trunk);
fp=fopen(fname_tsv,'w');
fprintf(fp,'%s\n',I_VariableName_{52:end});
fclose(fp);
%%%%%%%%;
if flag_E;
fname_tsv = sprintf('%s/dir_mat/E_CLabel_pair_.nsv',dir_trunk);
fp=fopen(fname_tsv,'w');
fprintf(fp,'%s\n',AB_E_.CLabel_pair_{1:end});
fclose(fp);
end;%if flag_E;
%%%%%%%%;
if flag_I;
fname_tsv = sprintf('%s/dir_mat/I_CLabel_pair_.nsv',dir_trunk);
fp=fopen(fname_tsv,'w');
fprintf(fp,'%s\n',AB_I_.CLabel_pair_{1:end});
fclose(fp);
end;%if flag_I;
%%%%%%%%;
if flag_E;
AB_E_AB_C_absZ_ = AB_E_.AB_C_absZ_;
fname_tsv = sprintf('%s/dir_mat/E_AB_C_absZ_.tsv',dir_trunk);
save(fname_tsv,'AB_E_AB_C_absZ_','-ascii','-tabs');
AB_E_AB_K_absZ_ = AB_E_.AB_K_absZ_;
fname_tsv = sprintf('%s/dir_mat/E_AB_K_absZ_.tsv',dir_trunk);
save(fname_tsv,'AB_E_AB_K_absZ_','-ascii','-tabs');
AB_E_AB_Z_absZ_ = AB_E_.AB_Z_absZ_;
fname_tsv = sprintf('%s/dir_mat/E_AB_Z_absZ_.tsv',dir_trunk);
save(fname_tsv,'AB_E_AB_Z_absZ_','-ascii','-tabs');
AB_E_AB_H_absZ_ = AB_E_.AB_H_absZ_;
fname_tsv = sprintf('%s/dir_mat/E_AB_H_absZ_.tsv',dir_trunk);
save(fname_tsv,'AB_E_AB_H_absZ_','-ascii','-tabs');
end;%if flag_E;
%%%%%%%%;
if flag_I;
AB_I_AB_C_absZ_ = AB_I_.AB_C_absZ_;
fname_tsv = sprintf('%s/dir_mat/I_AB_C_absZ_.tsv',dir_trunk);
save(fname_tsv,'AB_I_AB_C_absZ_','-ascii','-tabs');
AB_I_AB_K_absZ_ = AB_I_.AB_K_absZ_;
fname_tsv = sprintf('%s/dir_mat/I_AB_K_absZ_.tsv',dir_trunk);
save(fname_tsv,'AB_I_AB_K_absZ_','-ascii','-tabs');
AB_I_AB_Z_absZ_ = AB_I_.AB_Z_absZ_;
fname_tsv = sprintf('%s/dir_mat/I_AB_Z_absZ_.tsv',dir_trunk);
save(fname_tsv,'AB_I_AB_Z_absZ_','-ascii','-tabs');
AB_I_AB_H_absZ_ = AB_I_.AB_H_absZ_;
fname_tsv = sprintf('%s/dir_mat/I_AB_H_absZ_.tsv',dir_trunk);
save(fname_tsv,'AB_I_AB_H_absZ_','-ascii','-tabs');
end;%if flag_I;
%%%%%%%%;
if flag_E;
tmp_ = E_rank_;
fname_tsv = sprintf('%s/dir_mat/E_rank_.tsv',dir_trunk);
save(fname_tsv,'tmp_','-ascii','-tabs');
tmp_ = zeta_E_un_(:,1:n_zeta_E);
fname_tsv = sprintf('%s/dir_mat/zeta_E_un_.tsv',dir_trunk);
save(fname_tsv,'tmp_','-ascii','-tabs');
tmp_ = zeta_E_vn_(:,1:n_zeta_E);
fname_tsv = sprintf('%s/dir_mat/zeta_E_vn_.tsv',dir_trunk);
save(fname_tsv,'tmp_','-ascii','-tabs');
tmp_ = beta_E_un_(:,1:n_beta_E);
fname_tsv = sprintf('%s/dir_mat/beta_E_un_.tsv',dir_trunk);
save(fname_tsv,'tmp_','-ascii','-tabs');
tmp_ = beta_E_vn_(:,1:n_beta_E);
fname_tsv = sprintf('%s/dir_mat/beta_E_vn_.tsv',dir_trunk);
save(fname_tsv,'tmp_','-ascii','-tabs');
end;%if flag_E;
%%%%%%%%;
if flag_I;
tmp_ = I_rank_;
fname_tsv = sprintf('%s/dir_mat/I_rank_.tsv',dir_trunk);
save(fname_tsv,'tmp_','-ascii','-tabs');
tmp_ = zeta_I_un_(:,1:n_zeta_I);
fname_tsv = sprintf('%s/dir_mat/zeta_I_un_.tsv',dir_trunk);
save(fname_tsv,'tmp_','-ascii','-tabs');
tmp_ = zeta_I_vn_(:,1:n_zeta_I);
fname_tsv = sprintf('%s/dir_mat/zeta_I_vn_.tsv',dir_trunk);
save(fname_tsv,'tmp_','-ascii','-tabs');
tmp_ = beta_I_un_(:,1:n_beta_I);
fname_tsv = sprintf('%s/dir_mat/beta_I_un_.tsv',dir_trunk);
save(fname_tsv,'tmp_','-ascii','-tabs');
tmp_ = beta_I_vn_(:,1:n_beta_I);
fname_tsv = sprintf('%s/dir_mat/beta_I_vn_.tsv',dir_trunk);
save(fname_tsv,'tmp_','-ascii','-tabs');
end;%if flag_I;
%%%%%%%%;

%%%%%%%
if flag_I; 
fname_mat = sprintf('%s/dir_mat/AB_I_.AB_X_absZ_.mat',dir_trunk); 
AB_I_n_u = AB_I_.n_u;
AB_I_n_CLabel = AB_I_.n_CLabel;
AB_I_n_CCOV = AB_I_.n_CCOV;
AB_I_n_Z_GENE = AB_I_.n_Z_GENE;
AB_I_CLabel_pair_ = AB_I_.CLabel_pair_;
AB_I_AB_C_absZ_ = AB_I_.AB_C_absZ_;
AB_I_AB_K_absZ_ = AB_I_.AB_K_absZ_;
AB_I_AB_Z_absZ_ = AB_I_.AB_Z_absZ_;
AB_I_AB_H_absZ_ = AB_I_.AB_H_absZ_;
disp(sprintf(' %% writing %s',fname_mat));
%save(fname_mat,'AB_I_n_u','AB_I_n_CLabel','AB_I_n_CCOV','AB_I_n_Z_GENE','AB_I_CLabel_pair_','AB_I_AB_C_absZ_','AB_I_AB_K_absZ_','AB_I_AB_Z_absZ_','AB_I_AB_H_absZ_')
save(fname_mat,'AB_I_n_u','AB_I_n_CLabel','AB_I_n_CCOV','AB_I_n_Z_GENE','AB_I_CLabel_pair_','AB_I_AB_C_absZ_','AB_I_AB_K_absZ_')

end;%if flag_I; 
%%%%%%%%;
if flag_E; 
fname_mat = sprintf('%s/dir_mat/AB_E_.AB_X_absZ_.mat',dir_trunk); 
AB_E_n_u = AB_E_.n_u;
AB_E_n_CLabel = AB_E_.n_CLabel;
AB_E_n_CCOV = AB_E_.n_CCOV;
AB_E_n_Z_GENE = AB_E_.n_Z_GENE;
AB_E_CLabel_pair_ = AB_E_.CLabel_pair_;
AB_E_AB_C_absZ_ = AB_E_.AB_C_absZ_;
AB_E_AB_K_absZ_ = AB_E_.AB_K_absZ_;
AB_E_AB_Z_absZ_ = AB_E_.AB_Z_absZ_;
AB_E_AB_H_absZ_ = AB_E_.AB_H_absZ_;
disp(sprintf(' %% writing %s',fname_mat));
%save(fname_mat,'AB_E_n_u','AB_E_n_CLabel','AB_E_n_CCOV','AB_E_n_Z_GENE','AB_E_CLabel_pair_','AB_E_AB_C_absZ_','AB_E_AB_K_absZ_','AB_E_AB_Z_absZ_','AB_E_AB_H_absZ_')
save(fname_mat,'AB_E_n_u','AB_E_n_CLabel','AB_E_n_CCOV','AB_E_n_Z_GENE','AB_E_CLabel_pair_','AB_E_AB_C_absZ_','AB_E_AB_K_absZ_')
end;%if flag_E; 
%%%%%%%%;


flag_test=1;
if flag_test;
%%%%%%%;
% Now we can test a pair of labels: ;
%%%%%%%%;
for nl=1:3;
if nl==1; nA=find(strcmp(u_CLabel_,'28')); nB=find(strcmp(u_CLabel_,'33')); end;
if nl==2; nA=find(strcmp(u_CLabel_,'22')); nB=find(strcmp(u_CLabel_,'11')); end;
if nl==3; nA=find(strcmp(u_CLabel_,'19')); nB=find(strcmp(u_CLabel_,'15')); end;
disp(sprintf(' %% Comparing Label %d (%s) and %d (%s):',nA,u_CLabel_{nA},nB,u_CLabel_{nB}));
%%%%%%%%;
if flag_I; test_loader_plot_sort_list_0(AB_I_,nA,nB,u_CLabel_,C_VariableName_LC_,dir_trunk,'I'); end;%if flag_I;
if flag_E; test_loader_plot_sort_list_0(AB_E_,nA,nB,u_CLabel_,C_VariableName_LC_,dir_trunk,'E'); end;%if flag_E;
%%%%%%%%;
end;%for nl=1:3;
end;%if flag_test;

disp('returning');return;






